File: L14GoldmanSachsMedium/FractionToRecurringDecimal.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <cmath>

using namespace std;

string fractionToDecimal(int numerator, int denominator)
{
    if (numerator == 0)
        return "0"; // If numerator is 0, result is "0"

    string result;

    // Determine the sign of the result
    if ((numerator < 0) ^ (denominator < 0))
    {
        result += "-";
    }

    // Convert numerator and denominator to long long to handle overflow
    long long num = abs((long long)numerator);
    long long den = abs((long long)denominator);

    // Compute the integer part of the result
    result += to_string(num / den);
    long long remainder = num % den;

    if (remainder == 0)
    {
        return result; // If no remainder, return the integer result
    }

    // Otherwise, process the fractional part
    result += ".";
    unordered_map<long long, int> remainderMap; // Store remainder positions

    while (remainder != 0)
    {
        if (remainderMap.find(remainder) != remainderMap.end())
        {
            // If remainder repeats, insert parentheses at the start of repetition
            result.insert(remainderMap[remainder], "(");
            result += ")";
            break;
        }

        // Store the position of this remainder
        remainderMap[remainder] = result.size();

        // Multiply remainder by 10 and append next digit
        remainder *= 10;
        result += to_string(remainder / den);
        remainder %= den;
    }

    return result;
}

int main()
{
    cout << fractionToDecimal(1, 2) << endl;   // Output: "0.5"
    cout << fractionToDecimal(2, 3) << endl;   // Output: "0.(6)"
    cout << fractionToDecimal(4, 333) << endl; // Output: "0.(012)"
    cout << fractionToDecimal(-50, 8) << endl; // Output: "-6.25"
    return 0;
}

File: L14GoldmanSachsMedium/FractionToRecurringDecimal.java

import java.util.*;

public class FractionToRecurringDecimal {

    public static String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) {
            return "0"; // If numerator is 0, result is 0
        }
        StringBuilder result = new StringBuilder();

        // Determine the sign of the result
        if ((numerator < 0) ^ (denominator < 0)) {
            result.append("-");
        }

        // Convert numerator and denominator to long to handle overflow
        long num = Math.abs((long) numerator);
        long den = Math.abs((long) denominator);

        // Compute the integer part of the result
        result.append(num / den);
        long remainder = num % den;

        if (remainder == 0) {
            return result.toString(); // If no remainder, return the integer result
        }

        // Otherwise, process the fractional part
        result.append(".");
        Map<Long, Integer> remainderMap = new HashMap<>(); // Store remainder positions

        while (remainder != 0) {
            if (remainderMap.containsKey(remainder)) {
                // If remainder repeats, insert parentheses at the start of repetition
                result.insert(remainderMap.get(remainder), "(");
                result.append(")");
                break;
            }

            // Store the position of this remainder
            remainderMap.put(remainder, result.length());

            // Multiply remainder by 10 and append next digit
            remainder *= 10;
            result.append(remainder / den);
            remainder %= den;
        }

        return result.toString();
    }

    public static void main(String[] args) {
        System.out.println(fractionToDecimal(1, 2));   // Output: "0.5"
        System.out.println(fractionToDecimal(2, 3));   // Output: "0.(6)"
        System.out.println(fractionToDecimal(4, 333)); // Output: "0.(012)"
        System.out.println(fractionToDecimal(-50, 8)); // Output: "-6.25"
    }
}
